gsql could not login to proxy if backend database is MogDB (openGauss) since SQL_ASCII charset

## Bug Report

### Which version of ShardingSphere did you use?
master branch, commit 0aa073b7921a2c761bc9fa9b6bcb9bfe03bf210f

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy, backend database is MogDB (based on openGauss)
macOS

### Expected behavior
Login successfully with gsql.
Access macOS host proxy from docker.

### Actual behavior
Client got error message:
```
$ gsql --host=host.docker.internal --port=3307 scaling_db root
gsql: ERROR:  SQL_ASCII
```

There's exception in proxy log:
```
[ERROR] 2022-04-09 22:02:47.501 [nioEventLoopGroup-3-1] o.a.s.p.f.n.FrontendChannelInboundHandler - Exception occur: 
java.nio.charset.UnsupportedCharsetException: SQL_ASCII
	at java.nio.charset.Charset.forName(Charset.java:531)
	at org.apache.shardingsphere.proxy.frontend.opengauss.authentication.OpenGaussAuthenticationEngine.processStartupMessage(OpenGaussAuthenticationEngine.java:105)
	at org.apache.shardingsphere.proxy.frontend.opengauss.authentication.OpenGaussAuthenticationEngine.authenticate(OpenGaussAuthenticationEngine.java:98)
	at org.apache.shardingsphere.proxy.frontend.netty.FrontendChannelInboundHandler.authenticate(FrontendChannelInboundHandler.java:83)
	at org.apache.shardingsphere.proxy.frontend.netty.FrontendChannelInboundHandler.channelRead(FrontendChannelInboundHandler.java:75)
```

### Reason analyze (If you can)

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

### Example codes for reproduce this issue (such as a github link).

--- a/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/PostgreSQLCharacterSets.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/PostgreSQLCharacterSets.java
@@ -1,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.proxy.backend.text.admin.postgresql;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.charset.UnsupportedCharsetException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * Character sets defined in PostgreSQL.
+ * <a href="https://www.postgresql.org/docs/14/multibyte.html">24.3. Character Set Support</a>
+ */
+public enum PostgreSQLCharacterSets {
+    
+    SQL_ASCII(() -> StandardCharsets.US_ASCII),
+    EUC_JP(() -> Charset.forName("EUC_JP")),
+    EUC_CN(() -> Charset.forName("EUC_CN")),
+    EUC_KR(() -> Charset.forName("EUC_KR")),
+    EUC_TW(() -> Charset.forName("EUC_TW")),
+    EUC_JIS_2004(() -> Charset.forName("EUC_JIS_2004")),
+    UTF8(() -> StandardCharsets.UTF_8, "Unicode", "UTF-8"),
+    MULE_INTERNAL(() -> Charset.forName("MULE_INTERNAL")),
+    LATIN1(() -> StandardCharsets.ISO_8859_1, "ISO88591"),
+    LATIN2(() -> Charset.forName("LATIN2"), "ISO88592"),
+    LATIN3(() -> Charset.forName("LATIN3"), "ISO88593"),
+    LATIN4(() -> Charset.forName("LATIN4"), "ISO88594"),
+    LATIN5(() -> Charset.forName("LATIN5"), "ISO88599"),
+    LATIN6(() -> Charset.forName("ISO-8859-10"), "ISO885910"),
+    LATIN7(() -> Charset.forName("ISO-8859-13"), "ISO885913"),
+    LATIN8(() -> Charset.forName("ISO-8859-14"), "ISO885914"),
+    LATIN9(() -> Charset.forName("LATIN9"), "ISO885915"),
+    LATIN10(() -> Charset.forName("LATIN10"), "ISO885916"),
+    WIN1256(() -> Charset.forName("WINDOWS-1256")),
+    WIN1258(() -> Charset.forName("WINDOWS-1258"), "ABC", "TCVN", "TCVN5712", "VSCII"),
+    WIN866(() -> Charset.forName("WINDOWS-866"), "ALT"),
+    WIN874(() -> Charset.forName("WINDOWS-874")),
+    KOI8R(() -> Charset.forName("KOI8-R")),
+    WIN1251(() -> Charset.forName("WINDOWS-1251"), "WIN"),
+    WIN1252(() -> Charset.forName("WINDOWS-1252")),
+    ISO_8859_5(() -> Charset.forName("ISO-8859-5")),
+    ISO_8859_6(() -> Charset.forName("ISO-8859-6")),
+    ISO_8859_7(() -> Charset.forName("ISO-8859-7")),
+    ISO_8859_8(() -> Charset.forName("ISO-8859-8")),
+    WIN1250(() -> Charset.forName("WINDOWS-1250")),
+    WIN1253(() -> Charset.forName("WINDOWS-1253")),
+    WIN1254(() -> Charset.forName("WINDOWS-1254")),
+    WIN1255(() -> Charset.forName("WINDOWS-1255")),
+    WIN1257(() -> Charset.forName("WINDOWS-1257")),
+    KOI8U(() -> Charset.forName("KOI8-U"), "KOI8"),
+    SJIS(() -> Charset.forName("SJIS"), "Mskanji", "ShiftJIS", "WIN932", "Windows932"),
+    BIG5(() -> Charset.forName("BIG5"), "WIN950", "Windows950"),
+    GBK(() -> Charset.forName("GBK"), "WIN936", "Windows936"),
+    UHC(() -> Charset.forName("UHC"), "WIN949", "Windows949"),
+    GB18030(() -> Charset.forName("GB18030")),
+    JOHAB(() -> Charset.forName("JOHAB")),
+    SHIFT_JIS_2004(() -> Charset.forName("SHIFT_JIS"));
+    
+    private static final Map<String, PostgreSQLCharacterSets> CHARACTER_SETS_MAP;
+    
+    static {
+        Map<String, PostgreSQLCharacterSets> map = new HashMap<>();
+        for (PostgreSQLCharacterSets each : values()) {
+            map.put(each.name(), each);
+            for (String eachAlias : each.aliases) {
+                map.put(eachAlias.toUpperCase(), each);
+            }
+        }
+        CHARACTER_SETS_MAP = map;
+    }
+    
+    private final Charset charset;
+    
+    private final String[] aliases;
+    
+    PostgreSQLCharacterSets(final Supplier<Charset> charsetSupplier, final String... aliases) {
+        Charset result = null;
+        try {
+            result = charsetSupplier.get();
+        } catch (UnsupportedCharsetException ignored) {
+        }
+        charset = result;
+        this.aliases = aliases;
+    }
+    
+    /**
+     * Find corresponding {@link Charset} by charset name defined in PostgreSQL.
+     *
+     * @param charsetName charset name defined in PostgreSQL
+     * @return corresponding {@link Charset}
+     */
+    public static Charset findCharacterSet(final String charsetName) {
+        PostgreSQLCharacterSets result = CHARACTER_SETS_MAP.get(charsetName.toUpperCase());
+        return null != result && null != result.charset ? result.charset : Charset.forName(charsetName);
+    }
+}

--- a/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/PostgreSQLCharacterSets.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/PostgreSQLCharacterSets.java
@@ -36,7 +36,7 @@ public enum PostgreSQLCharacterSets {
     EUC_KR(() -> Charset.forName("EUC_KR")),
     EUC_TW(() -> Charset.forName("EUC_TW")),
     EUC_JIS_2004(() -> Charset.forName("EUC_JIS_2004")),
-    UTF8(() -> StandardCharsets.UTF_8, "Unicode", "UTF-8"),
+    UTF8(() -> StandardCharsets.UTF_8, "Unicode"),
     MULE_INTERNAL(() -> Charset.forName("MULE_INTERNAL")),
     LATIN1(() -> StandardCharsets.ISO_8859_1, "ISO88591"),
     LATIN2(() -> Charset.forName("LATIN2"), "ISO88592"),
@@ -76,7 +76,7 @@ public enum PostgreSQLCharacterSets {
     private static final Map<String, PostgreSQLCharacterSets> CHARACTER_SETS_MAP;
     
     static {
-        Map<String, PostgreSQLCharacterSets> map = new HashMap<>();
+        Map<String, PostgreSQLCharacterSets> map = new HashMap<>(128);
         for (PostgreSQLCharacterSets each : values()) {
             map.put(each.name(), each);
             for (String eachAlias : each.aliases) {

--- a/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/executor/PostgreSQLSetCharsetExecutor.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/executor/PostgreSQLSetCharsetExecutor.java
@@ -21,6 +21,7 @@ import lombok.RequiredArgsConstructor;
 import org.apache.shardingsphere.db.protocol.CommonConstants;
 import org.apache.shardingsphere.proxy.backend.session.ConnectionSession;
 import org.apache.shardingsphere.proxy.backend.text.admin.executor.DatabaseSetCharsetExecutor;
+import org.apache.shardingsphere.proxy.backend.text.admin.postgresql.PostgreSQLCharacterSets;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dal.VariableAssignSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.statement.dal.SetStatement;
 
@@ -54,13 +55,7 @@ public final class PostgreSQLSetCharsetExecutor implements DatabaseSetCharsetExe
     
     private Charset parseCharset(final String value) {
         String result = value.toLowerCase(Locale.ROOT);
-        switch (result) {
-            case "default":
-            case "unicode":
-                return StandardCharsets.UTF_8;
-            default:
-                return Charset.forName(value);
-        }
+        return "default".equals(result) ? StandardCharsets.UTF_8 : PostgreSQLCharacterSets.findCharacterSet(result);
     }
     
     @Override

--- a/shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/PostgreSQLCharacterSetsTest.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-backend/src/test/java/org/apache/shardingsphere/proxy/backend/text/admin/postgresql/PostgreSQLCharacterSetsTest.java
@@ -1,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.proxy.backend.text.admin.postgresql;
+
+import org.junit.Test;
+
+import java.nio.charset.StandardCharsets;
+import java.nio.charset.UnsupportedCharsetException;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+public final class PostgreSQLCharacterSetsTest {
+    
+    @Test
+    public void assertFindSqlAscii() {
+        assertThat(PostgreSQLCharacterSets.findCharacterSet("SQL_ASCII"), is(StandardCharsets.US_ASCII));
+    }
+    
+    @Test
+    public void assertFindUTF8() {
+        assertThat(PostgreSQLCharacterSets.findCharacterSet("utf8"), is(StandardCharsets.UTF_8));
+    }
+    
+    @Test(expected = UnsupportedCharsetException.class)
+    public void assertFindUnsupportedCharset() {
+        PostgreSQLCharacterSets.findCharacterSet("unknown_charset");
+    }
+}

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngine.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngine.java
@@ -22,6 +22,7 @@ import io.netty.channel.ChannelHandlerContext;
 import org.apache.shardingsphere.db.protocol.CommonConstants;
 import org.apache.shardingsphere.db.protocol.opengauss.packet.authentication.OpenGaussAuthenticationSCRAMSha256Packet;
 import org.apache.shardingsphere.db.protocol.payload.PacketPayload;
+import org.apache.shardingsphere.proxy.backend.text.admin.postgresql.PostgreSQLCharacterSets;
 import org.apache.shardingsphere.db.protocol.postgresql.constant.PostgreSQLErrorCode;
 import org.apache.shardingsphere.db.protocol.postgresql.constant.PostgreSQLServerInfo;
 import org.apache.shardingsphere.db.protocol.postgresql.packet.command.query.extended.PostgreSQLPreparedStatementRegistry;
@@ -42,7 +43,6 @@ import org.apache.shardingsphere.proxy.frontend.postgresql.authentication.except
 import org.apache.shardingsphere.proxy.frontend.postgresql.authentication.exception.PostgreSQLAuthenticationException;
 import org.apache.shardingsphere.proxy.frontend.postgresql.authentication.exception.PostgreSQLProtocolViolationException;
 
-import java.nio.charset.Charset;
 import java.util.concurrent.ThreadLocalRandom;
 
 /**
@@ -102,7 +102,7 @@ public final class OpenGaussAuthenticationEngine implements AuthenticationEngine
         startupMessageReceived = true;
         PostgreSQLComStartupPacket comStartupPacket = new PostgreSQLComStartupPacket(payload);
         clientEncoding = comStartupPacket.getClientEncoding();
-        context.channel().attr(CommonConstants.CHARSET_ATTRIBUTE_KEY).set(Charset.forName(clientEncoding));
+        context.channel().attr(CommonConstants.CHARSET_ATTRIBUTE_KEY).set(PostgreSQLCharacterSets.findCharacterSet(clientEncoding));
         String user = comStartupPacket.getUser();
         if (Strings.isNullOrEmpty(user)) {
             throw new InvalidAuthorizationSpecificationException("no PostgreSQL user name specified in startup packet");

--- a/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/authentication/PostgreSQLAuthenticationEngine.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-postgresql/src/main/java/org/apache/shardingsphere/proxy/frontend/postgresql/authentication/PostgreSQLAuthenticationEngine.java
@@ -22,6 +22,7 @@ import io.netty.channel.ChannelHandlerContext;
 import org.apache.shardingsphere.db.protocol.CommonConstants;
 import org.apache.shardingsphere.db.protocol.payload.PacketPayload;
 import org.apache.shardingsphere.db.protocol.postgresql.constant.PostgreSQLAuthenticationMethod;
+import org.apache.shardingsphere.proxy.backend.text.admin.postgresql.PostgreSQLCharacterSets;
 import org.apache.shardingsphere.db.protocol.postgresql.constant.PostgreSQLErrorCode;
 import org.apache.shardingsphere.db.protocol.postgresql.constant.PostgreSQLServerInfo;
 import org.apache.shardingsphere.db.protocol.postgresql.packet.command.query.extended.PostgreSQLPreparedStatementRegistry;
@@ -45,8 +46,6 @@ import org.apache.shardingsphere.proxy.frontend.postgresql.authentication.authen
 import org.apache.shardingsphere.proxy.frontend.postgresql.authentication.exception.InvalidAuthorizationSpecificationException;
 import org.apache.shardingsphere.proxy.frontend.postgresql.authentication.exception.PostgreSQLAuthenticationException;
 import org.apache.shardingsphere.proxy.frontend.postgresql.authentication.exception.PostgreSQLProtocolViolationException;
-
-import java.nio.charset.Charset;
 
 /**
  * Authentication engine for PostgreSQL.
@@ -88,7 +87,7 @@ public final class PostgreSQLAuthenticationEngine implements AuthenticationEngin
         startupMessageReceived = true;
         PostgreSQLComStartupPacket comStartupPacket = new PostgreSQLComStartupPacket(payload);
         clientEncoding = comStartupPacket.getClientEncoding();
-        context.channel().attr(CommonConstants.CHARSET_ATTRIBUTE_KEY).set(Charset.forName(clientEncoding));
+        context.channel().attr(CommonConstants.CHARSET_ATTRIBUTE_KEY).set(PostgreSQLCharacterSets.findCharacterSet(clientEncoding));
         String user = comStartupPacket.getUser();
         if (Strings.isNullOrEmpty(user)) {
             throw new InvalidAuthorizationSpecificationException("no PostgreSQL user name specified in startup packet");

